### PR TITLE
v5

### PR DESCRIPTION
**Pull request by manovotny**
The list below will be updated as new commits are made to the PR.

---

Breaking

-   Requires eslint `6.2.1` or newer.

Added

-   Adds rules.
    -   [array-element-newline](https://github.com/eslint/eslint/blob/master/docs/rules/array-element-newline.md)
    -   [eslint-comments/no-unused-enable](https://github.com/mysticatea/eslint-plugin-eslint-comments/blob/master/docs/rules/no-unused-enable.md)
    -   [function-call-argument-newline](https://github.com/eslint/eslint/blob/master/docs/rules/function-call-argument-newline.md)
    -   [jest/no-alias-methods](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/no-alias-methods.md)
    -   [jest/no-commented-out-tests](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/no-commented-out-tests.md)
    -   [jest/no-duplicate-hooks](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/no-duplicate-hooks.md)
    -   [jest/no-empty-title](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/no-empty-title.md)
    -   [jest/no-expect-resolves](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/no-expect-resolves.md)
    -   [jest/no-if](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/no-if.md)
    -   [jest/no-jest-import](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/no-jest-import.md)
    -   [jest/no-mocks-import](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/no-mocks-import.md)
    -   [jest/no-standalone-expect](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/no-standalone-expect.md)
    -   [jest/no-test-callback](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/no-test-callback.md)
    -   [jest/no-test-prefixes](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/no-test-prefixes.md)
    -   [jest/no-test-return-statement](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/no-test-return-statement.md)
    -   [jest/no-truthy-falsy](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/no-truthy-falsy.md)
    -   [jest/no-try-expect](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/no-try-expect.md)
    -   [jest/prefer-called-with](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/prefer-called-with.md)
    -   [jest/prefer-spy-on](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/prefer-spy-on.md)
    -   [jest/prefer-strict-equal](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/prefer-strict-equal.md)
    -   [jest/prefer-to-contain](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/prefer-to-contain.md)
    -   [jest/prefer-todo](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/prefer-todo.md)
    -   [jest/require-tothrow-message](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/require-tothrow-message.md)
    -   [max-classes-per-file](https://github.com/eslint/eslint/blob/master/docs/rules/max-classes-per-file.md)
    -   [no-async-promise-executor](https://github.com/eslint/eslint/blob/master/docs/rules/no-async-promise-executor.md)
    -   [no-await-in-loop](https://github.com/eslint/eslint/blob/master/docs/rules/no-await-in-loop.md)
    -   [no-misleading-character-class](https://github.com/eslint/eslint/blob/master/docs/rules/no-misleading-character-class.md)
    -   [no-useless-catch](https://github.com/eslint/eslint/blob/master/docs/rules/no-useless-catch.md)
    -   [node/no-unsupported-features/es-builtins](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/no-unsupported-features/es-builtins.md)
    -   [node/no-unsupported-features/node-builtins](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/no-unsupported-features/node-builtins.md)
    -   [node/prefer-global/buffer](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/prefer-global/buffer.md)
    -   [node/prefer-global/console](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/prefer-global/console.md)
    -   [node/prefer-global/process](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/prefer-global/process.md)
    -   [node/prefer-global/text-decoder](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/prefer-global/text-decoder.md)
    -   [node/prefer-global/text-encoder](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/prefer-global/text-encoder.md)
    -   [node/prefer-global/url](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/prefer-global/url.md)
    -   [node/prefer-global/url-search-params](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/prefer-global/url-search-params.md)
    -   [prefer-object-spread](https://github.com/eslint/eslint/blob/master/docs/rules/prefer-object-spread.md)
    -   [react/jsx-curly-newline](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-curly-newline.md)
    -   [react/jsx-fragments](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-fragments.md)
    -   [react/jsx-one-expression-per-line](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-one-expression-per-line.md)
    -   [react/jsx-props-no-multi-spaces](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-props-no-multi-spaces.md)
    -   [react/no-unsafe](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-unsafe.md)
    -   [react/prefer-read-only-props](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prefer-read-only-props.md)
    -   [react/state-in-constructor](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/state-in-constructor.md)
    -   [react/static-property-placement](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/static-property-placement.md)
    -   react-hooks/exhaustive-deps
    -   react-hooks/rules-of-hooks
    -   [react-native/no-unused-styles](https://github.com/Intellicode/eslint-plugin-react-native/blob/master/docs/rules/no-unused-styles.md)
    -   [react-native/sort-styles](https://github.com/Intellicode/eslint-plugin-react-native/blob/master/docs/rules/sort-styles.md)
    -   [require-unicode-regexp](https://github.com/eslint/eslint/blob/master/docs/rules/require-unicode-regexp.md)
    -   [unicorn/filename-case](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/filename-case.md)
    -   [unicorn/no-abusive-eslint-disable](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/no-abusive-eslint-disable.md)
    -   [unicorn/no-console-spaces](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/no-console-spaces.md)
    -   [unicorn/no-for-loop](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/no-for-loop.md)
    -   [unicorn/no-unused-properties](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/no-unused-properties.md)
    -   [unicorn/no-zero-fractions](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/no-zero-fractions.md)
    -   [unicorn/prefer-event-key](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/prefer-event-key.md)
    -   [unicorn/prefer-flat-map](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/prefer-flat-map.md)
    -   [unicorn/prefer-includes](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/prefer-includes.md)
    -   [unicorn/prefer-node-append](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/prefer-node-append.md)
    -   [unicorn/prefer-node-remove](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/prefer-node-remove.md)
    -   [unicorn/prefer-query-selector](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/prefer-query-selector.md)
    -   [unicorn/prefer-text-content](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/prefer-text-content.md)
    -   [unicorn/regex-shorthand](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/regex-shorthand.md)

Changed

-   Dynamically enables / disables plugins (and thus rules) based on `package.json` dependencies.
-   Changes `ecmaVersion` from `2019` to `2020`.
-   Changes rule options.
    -   [camelcase](https://github.com/eslint/eslint/blob/master/docs/rules/camelcase.md)
        -   Adds `ignoreDestructuring` option.
    -   [import/no-useless-path-segments](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-useless-path-segments.md)
        -   Adds `noUselessIndex` option.
    -   [jest/valid-expect](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/valid-expect.md)
        -   Adds `alwaysAwait: true` option.
    -   [no-unused-vars](https://github.com/eslint/eslint/blob/master/docs/rules/no-unused-vars.md)
        -   Adds `args`, `caughtErrors`, `ignoreRestSiblings`, and `vars` options.
    -   [react/jsx-indent](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-indent.md)
        -   Adds `checkAttributes` and `indentLogicalExpressions` options.
    -   [react/no-string-refs](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-string-refs.md)
        -   Adds `noTemplateLiterals` option.
    -   [unicorn/filename-case](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/filename-case.md)
        -   Adds `pascalCase` option.

Removed

-   Removes rules.
    -   [complexity](https://github.com/eslint/eslint/blob/master/docs/rules/complexity.md)
    -   [max-depth](https://github.com/eslint/eslint/blob/master/docs/rules/max-depth.md)
    -   [max-nested-callbacks](https://github.com/eslint/eslint/blob/master/docs/rules/max-nested-callbacks.md)
    -   [max-params](https://github.com/eslint/eslint/blob/master/docs/rules/max-params.md)
    -   [no-catch-shadow](https://github.com/eslint/eslint/blob/master/docs/rules/no-catch-shadow.md)
    -   [one-var](https://github.com/eslint/eslint/blob/master/docs/rules/one-var.md)
    -   [import/no-dynamic-require](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-dynamic-require.md)
    -   [jsx-a11y/label-has-for](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/label-has-for.md)
    -   [prefer-object-spread/prefer-object-spread](https://github.com/bryanrsmith/eslint-plugin-prefer-object-spread)
    -   [react/no-multi-comp](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-multi-comp.md)
    -   [react/no-set-state](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-set-state.md)
    -   [react/prop-types](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prop-types.md)
    -   [unicorn/no-process-exit](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/no-process-exit.md)

Changeset

[`v4.0.1...v5.0.0`](https://github.com/manovotny/eslint-config-get-off-my-lawn/compare/v4.0.1...v5.0.0)

### CODE REVIEW

File: src/react.js [hunk 1]
```diff
@@ -110,12 +118,12 @@ module.exports = {
     ],
     'react/no-redundant-should-component-update': 'error',
     'react/no-render-return-value': 'error',
-    'react/no-set-state': 'error',
```
Comment (leerob) [inline src/react.js]: 🙌

File: src/import.js [hunk 2]
```diff
@@ -16,7 +16,6 @@ module.exports = {
     'import/no-amd': 'error',
     'import/no-deprecated': 'error',
     'import/no-duplicates': 'error',
-    'import/no-dynamic-require': 'error',
```
Comment (leerob) [inline src/import.js]: FWIW I agree with your reasoning to remove this.

File: src/react.js [hunk 3]
```diff
@@ -78,7 +79,14 @@ module.exports = {
             allowGlobals: true
         }
     ],
+    'react/jsx-one-expression-per-line': [
+        'error',
+        {
+            allow: 'single-child'
```
Comment (leerob) [inline src/react.js]: I'm assuming you're thinking this use case: 
```js
<App>Hello</App>
<App>{"Hello"}</App>
```

versus this one, which is meh. 

```js
<App><Hello /></App>
```
I don't know that I see this very often, though. Would Prettier take care of this anyway?

File: src/react.js [hunk 4]
```diff
@@ -59,6 +59,7 @@ module.exports = {
         }
     ],
     'react/jsx-first-prop-new-line': ['error', 'multiline'],
+    'react/jsx-fragments': ['error', 'syntax'],
```
Comment (leerob) [inline src/react.js]: YES

File: src/eslint.js [hunk 5]
```diff
@@ -14,7 +15,6 @@ module.exports = {
     'comma-dangle': 'error',
     'comma-spacing': 'error',
     'comma-style': 'error',
-    complexity: ['error', 4],
```
Comment (leerob) [inline src/eslint.js]: Did you decide it's just not worth fighting this battle? Or does this just fall back to the default?

File: index.js [hunk 6]
```diff
@@ -1,31 +1,44 @@
...
 const node = require('./src/node');
 const objects = require('./src/objects');
-const preferObjectSpread = require('./src/prefer-object-spread');
 const react = require('./src/react');
+const reactHooks = require('./src/react-hooks');
+const reactNative = require('./src/react-native');
 const security = require('./src/security');
 const unicorn = require('./src/unicorn');
 
+const getReactVesion = () => {
```
Comment (leerob) [inline index.js]: Is this used to determine whether to enforce hooks lints?

File: src/react.js (hunk 3 above)
Comment (manovotny) [reply src/react.js]: Actually, all three of the ones you listed would be allowed with this rule configured this way.

https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-one-expression-per-line.md#allow

And yes, prettier would handle this, but this repo doesn't work with prettier. That's why [eslint-config-get-off-my-lawn-prettier](https://github.com/manovotny/eslint-config-get-off-my-lawn-prettier) exists (which would disable this rule), though I have ideas on how to make it all work in a single repo now. 😉

File: src/eslint.js (hunk 5 above)
Comment (manovotny) [reply src/eslint.js]: No, it's completely disabled. You can [read this](https://github.com/manovotny/eslint-config-get-off-my-lawn/issues/95) for insight into why I disabled it.

File: index.js (hunk 6 above)
Comment (manovotny) [reply index.js]: No. [eslint-plugin-react recently added a `version` property](https://github.com/yannickcr/eslint-plugin-react#configuration). When it's not supplied, it defaults to `detect`, but if you use `get-off-my-lawn` without using React (like a pure node module, for example) and it cannot detect a version of React, then it throws warnings when running lint. This circumvents the warning by explicitly setting it, though I have a different idea on how to solve this problem now.

**Pull request merged**